File: api/cxx/pcbdataset.cpp

// API implementation.
// ~ Christopher Pramerdorfer, Computer Vision Lab, Vienna University of Technology

#include "pcbdataset.hpp"

#include <boost/foreach.hpp>
#include <boost/filesystem.hpp>
#include <boost/algorithm/string.hpp>

#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>

#include <cmath>
#include <stdexcept>
#include <utility>
#include <algorithm>
#include <fstream>

namespace pcbdataset
{
    Annot::Annot(const cv::RotatedRect& rect, float scale, const std::string& text) :
        rect(rect),
        scale(scale),
        text(text)
    { }

    float Annot::sizePixels(bool scaled) const
    {
        float sz = rect.size.area();
        if(! scaled)
            sz /= scale;

        return sz;
    }

    float Annot::sizeCm2(bool scaled) const
    {
        float sz = (rect.size.width / 87.4) * (rect.size.height / 87.4);
        if(! scaled)
            sz /= scale;

        return sz;
    }

    float Annot::aspect() const
    {
        return std::max(rect.size.width, rect.size.height) / std::min(rect.size.width, rect.size.height);
    }



    PCB::PCB(const std::string& root, float scale) :
        root(root),
        scale(scale)
    {
        if(! boost::filesystem::is_directory(root))
            throw std::invalid_argument("Root path is not a directory.");

        if(root.substr(root.size()-1, 1) == "/")
            throw std::invalid_argument("Root path must not end with /");

        if(scale <= 0 || scale > 2)
            throw std::invalid_argument("Scale must be > 0 and <= 2.");

        boost::filesystem::directory_iterator itr(root), endItr;
        BOOST_FOREACH(const boost::filesystem::path& p, std::make_pair(itr, endItr)) {
            if(boost::filesystem::is_regular_file(p)) {
                std::string fname = p.filename().string();
                if(fname.size() > 7 && fname.substr(0, 3) == "rec" && fname.substr(fname.size()-4, 4) == ".jpg") {
                    std::string rid = fname.substr(3, std::string::npos); rid = rid.substr(0, rid.size()-4);
                    _recordings.insert({ std::stoi(rid), p });
                }
            }
        }
    }

    int PCB::id() const
    {
        return std::stoi(root.filename().string().substr(3, std::string::npos));
    }

    std::vector<int> PCB::recordings() const
    {
        std::vector<int> ret;
        for(const auto& r : _recordings)
            ret.push_back(r.first);

        return ret;
    }

    cv::Mat PCB::image(int rec) const
    {
        if(_recordings.count(rec) == 0)
            throw std::invalid_argument("Recording does not exist.");

        cv::Mat image = cv::imread(_recordings.at(rec).string(), cv::IMREAD_UNCHANGED);
        if(image.rows == 0)
            throw std::runtime_error("Could not load the image.");

        if(scale != 1)
            cv::resize(image, image, cv::Size(0, 0), scale, scale);

        return image;
    }

    cv::Mat PCB::mask(int rec) const
    {
        if(_recordings.count(rec) == 0)
            throw std::invalid_argument("Recording does not exist.");

        boost::filesystem::path impath = _recordings.at(rec);
        std::stringstream ss; ss << "rec" << rec << "-mask.png";
        impath = impath.parent_path() / ss.str();

        if(! boost::filesystem::is_regular_file(impath))
            throw std::runtime_error("Mask file does not exist.");

        cv::Mat image = cv::imread(impath.string(), cv::IMREAD_GRAYSCALE);
        if(image.rows == 0)
            throw std::runtime_error("Could not load the image.");

        if(scale != 1)
            cv::resize(image, image, cv::Size(0, 0), scale, scale);

        return image;
    }

    cv::Mat PCB::imageMasked(int rec)
    {
        cv::Mat im = image(rec);
        cv::Mat msk = mask(rec);

        im.setTo(0, msk == 0);
        cv::Rect ci = _cropinfo(rec);

        return im(ci);
    }

    std::vector<Annot> PCB::ics(int rec, bool cropped, cv::Vec2f size, cv::Vec2f aspect)
    {
        if(_cache_ics.count(rec) == 1)
            return _cache_ics[rec];

        if(_recordings.count(rec) == 0)
            throw std::invalid_argument("Recording does not exist.");

        boost::filesystem::path fpath = _recordings.at(rec);
        std::stringstream ss; ss << "rec" << rec << "-annot.txt";
        fpath = fpath.parent_path() / ss.str();

        if(! boost::filesystem::is_regular_file(fpath))
            throw std::runtime_error("Annotation file does not exist.");

        std::vector<Annot> ret;

        std::string line;
        std::ifstream file(fpath.string());

        while(std::getline(file, line)) {
            std::vector<std::string> split;
            boost::split(split, line, boost::is_any_of(" "));

            if(split.size() < 5)
                throw std::runtime_error("Invalid line encountered while parsing file.");

            cv::RotatedRect rr(
                cv::Point2f(std::stof(split[0]), std::stof(split[1])),
                cv::Size2f(std::stof(split[2]), std::stof(split[3])),
                std::stof(split[4])
            );

            Annot tmp(rr, 1.0, "");

            if(size(0) > 0 && tmp.sizeCm2(false) < size(0))
                continue;

            if(size(1) > 0 && tmp.sizeCm2(false) > size(1))
                continue;

            if(aspect(0) > 0 && tmp.aspect() < aspect(0))
                continue;

            if(aspect(1) > 0 && tmp.aspect() > aspect(1))
                continue;

            if(scale != 1) {
                rr.center.x *= scale;
                rr.center.y *= scale;
                rr.size.width *= scale;
                rr.size.height *= scale;
            }

            if(cropped) {
                cv::Rect ci = _cropinfo(rec);
                rr.center.x -= ci.x;
                rr.center.y -= ci.y;
            }

            std::stringstream ss;
            for(size_t i = 5; i < split.size(); i++) {
                if(! split[i].empty()) {
                    ss << split[i];
                    if(i < split.size()-1)
                        ss << " ";
                }
            }

            ret.push_back(Annot(rr, scale, ss.str()));
        }

        file.close();

        _cache_ics.insert({ rec, std::move(ret) });

        return _cache_ics[rec];
    }

    cv::Rect PCB::_cropinfo(int rec)
    {
        if(_cache_cropinfo.count(rec) == 1)
            return _cache_cropinfo[rec];

        cv::Mat im = mask(rec);

        std::vector<std::vector<cv::Point> > cnt;
        cv::findContours(im, cnt, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_NONE);

        std::sort(cnt.begin(), cnt.end(), [](const std::vector<cv::Point>& a, const std::vector<cv::Point>& b) {
            cv::RotatedRect ra = cv::minAreaRect(a);
            cv::RotatedRect rb = cv::minAreaRect(b);

            return ra.size.area() < rb.size.area();
        });

        _cache_cropinfo.insert({ rec, cv::boundingRect(cnt[cnt.size()-1]) });

        return _cache_cropinfo[rec];
    }



    PCBDataset::PCBDataset(const std::string& root) : root(root)
    {
        if(! boost::filesystem::is_directory(root))
            throw std::invalid_argument("Root path is not a directory.");

        if(root.substr(root.size()-1, 1) == "/")
            throw std::invalid_argument("Root path must not end with /");

        boost::filesystem::directory_iterator itr(root), endItr;
        BOOST_FOREACH(const boost::filesystem::path& p, std::make_pair(itr, endItr)) {
            if(boost::filesystem::is_directory(p)) {
                std::string fname = p.filename().string();
                if(fname.size() > 3 && fname.substr(0, 3) == "pcb")
                    _pcbPaths.insert({ std::stoi(fname.substr(3, std::string::npos)), p });
            }
        }

        if(_pcbPaths.empty())
            throw std::runtime_error("Specified path contains no PCB directories.");
    }

    int PCBDataset::numPCBs() const
    {
        return _pcbPaths.size();
    }

    std::vector<int> PCBDataset::pcbIDs() const
    {
        std::vector<int> ret;
        for(const auto& p : _pcbPaths)
            ret.push_back(p.first);

        std::sort(ret.begin(), ret.end());

        return ret;
    }

    PCB PCBDataset::pcb(int id, float scale) const
    {
        if(_pcbPaths.count(id) == 0)
            throw std::invalid_argument("PCB does not exist.");

        return PCB(_pcbPaths.at(id).string(), scale);
    }
}

File: api/python/pcb_dataset.py

'''
Python2 API for the PCB DSLR dataset.
~ Christopher Pramerdorfer, Computer Vision Lab, Vienna University of Technology
'''

import cv2
import numpy as np

import os
import os.path
import argparse


class Annot:

    '''
    An annotated PCB component.
    '''

    def __init__(self, rect, scale, text=''):
        '''
        Constructor.
        rect: region of the component specified by ((cx, cy), (dx, dy), angle) as in OpenCV.
        scale: scale factor (affects size_pixels() and size_cm2()).
        text: optional label text
        '''

        self.rect = rect
        self._scale = scale
        self.text = text

    def __repr__(self):
        '''
        Returns a string representation.
        '''

        return 'Annotation {} "{}"'.format(self.rect, self.text)

    def size_pixels(self, scaled=True):
        '''
        Returns the size of the component in pixels.
        scaled: whether to regard the scale factor.
        '''

        sz = self.rect[1][0]*self.rect[1][1]
        if not scaled:
            sz /= self._scale

        return sz

    def size_cm2(self, scaled=True):
        '''
        Returns the size of the component in cm^2.
        scaled: whether to regard the scale factor.
        '''

        sz = (self.rect[1][0]/87.4) * (self.rect[1][1]/87.4)
        if not scaled:
            sz /= self._scale

        return sz

    def aspect(self):
        '''
        Returns the aspect ratio (larger side length / smaller side length).
        '''

        return max(self.rect[1][0], self.rect[1][1]) / min(self.rect[1][0], self.rect[1][1])


class PCB:

    '''
    A printed circuit board.
    '''

    def __init__(self, root, scale=1):
        '''
        Constructor.
        root: root directory path.
        scale: scale factor (1 = original size).
        '''

        if not os.path.isdir(root):
            raise Exception('Path "{}" is not a directory'.format(root))

        if scale <= 0 or scale > 2:
            raise Exception('Scale must be > 0 and <= 2')

        self._root = root
        self._scale = scale
        self._recordings = [int(os.path.splitext(p)[0][3:]) for p in os.listdir(root) if p.startswith('rec') and p.endswith('.jpg') and 'mask' not in p]
        self._cache_cropinfo = {}

    def __repr__(self):
        '''
        Returns a string representation.
        '''

        return 'PCB {} ({} recordings)'.format(self.id(), len(self._recordings))

    def id(self):
        '''
        Returns the PCB ID.
        '''

        return int(os.path.splitext(os.path.basename(self._root))[0][3:])

    def recordings(self):
        '''
        Returns a list of IDs of all available recordings.
        '''

        return self._recordings

    def image(self, rec=1):
        '''
        Returns the image of the specified recording.
        rec: desired recording (see recordings()).
        '''

        if rec not in self._recordings:
            raise Exception('Recording {} does not exist for this PCB'.format(rec))

        im = cv2.imread(os.path.join(self._root, 'rec{}.jpg'.format(rec)), cv2.IMREAD_UNCHANGED)
        if im.size == 0:
            raise Exception('Could not load the image')

        if self._scale != 1:
            im = cv2.resize(im, (0, 0), im, self._scale, self._scale)

        return im

    def mask(self, rec=1):
        '''
        Returns the mask of the specified recording.
        rec: desired recording (see recordings()).
        '''

        if rec not in self._recordings:
            raise Exception('Recording {} does not exist for this PCB'.format(rec))

        im = cv2.imread(os.path.join(self._root, 'rec{}-mask.png'.format(rec)), cv2.IMREAD_GRAYSCALE)
        if im.size == 0:
            raise Exception('Could not load the mask')

        if self._scale != 1:
            im = cv2.resize(im, (0, 0), im, self._scale, self._scale)

        return im

    def image_masked(self, rec=1):
        '''
        Returns the image of the specified recording, masked by the corresponding mask and cropped to remove background.
        rec: desired recording (see recordings()).
        '''

        im = self.image(rec)
        mask = self.mask(rec)

        im[mask == 0, :] = 0
        ci = self._cropinfo(rec)

        return im[ci[1]:ci[1]+ci[3], ci[0]:ci[0]+ci[2]]

    def ics(self, rec=1, cropped=False, size=(0, 0), aspect=(0, 0)):
        '''
        Returns a list of IC chips as a list of Annot objects.
        rec: desired recording (see recordings()).
        cropped: whether to return coordinates for cropped images (see image_masked()).
        size: (min, max) size of returned ICs in cm^2, disregarding the scale factor (0 = all).
        aspect: (min, max) aspect ratio of returned ICs (0 = all).
        '''

        if rec not in self._recordings:
            raise Exception('Recording {} does not exist for this PCB'.format(rec))

        fpath = os.path.join(self._root, 'rec{}-annot.txt'.format(rec))
        if not os.path.isfile(fpath):
            raise Exception('"{}" is not a file'.format(fpath))

        lines = None
        with open(fpath) as f:
            lines = [l.strip().split() for l in f.readlines()]

        ret = []
        for l in lines:
            l = [x.strip() for x in l]
            if len(l) < 5:
                raise Exception('Failed to parse line "{}"'.format(l))

            rect = [float(s) for s in l[:5]]
            text = '' if len(l) == 5 else ' '.join(l[5:])

            sz = (rect[2]/87.4, rect[3]/87.4)
            asp = max(sz[0], sz[1]) / min(sz[0], sz[1])
            sz = sz[0]*sz[1]

            if size[0] > 0 and sz < size[0]:
                continue

            if size[1] > 0 and sz > size[1]:
                continue

            if aspect[0] > 0 and asp < aspect[0]:
                continue

            if aspect[1] > 0 and asp > aspect[1]:
                continue

            if self._scale != 1:
                rect[0] *= self._scale
                rect[1] *= self._scale
                rect[2] *= self._scale
                rect[3] *= self._scale

            if cropped:
                ci = self._cropinfo(rec)
                rect[0] -= ci[0]
                rect[1] -= ci[1]

            ret.append(Annot((tuple(rect[0:2]), tuple(rect[2:4]), rect[4]), self._scale, text))

        return ret

    def _cropinfo(self, rec):
        '''
        Return (and cache) information for auto cropping a PCB image.
        rec: desired recording (see recordings()).
        '''

        if rec in self._cache_cropinfo:
            return self._cache_cropinfo[rec]

        im = self.mask(rec)
        cnt, _ = cv2.findContours(im, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)

        if len(cnt) > 1:  # use largest region if there are multiple
            def cntsz(c):
                rr = cv2.minAreaRect(c)
                return rr[1][0]*rr[1][1]

            cnt = sorted(cnt, key=cntsz, reverse=True)

        cx, cy, cw, ch = cv2.boundingRect(cnt[0])
        self._cache_cropinfo[rec] = (cx, cy, cw, ch)

        return self._cache_cropinfo[rec]


class PCBDataset:

    '''
    A PCB dataset.
    '''

    def __init__(self, root):
        '''
        Constructor.
        root: root path to the dataset.
        '''

        if not os.path.isdir(root):
            raise Exception('Path "{}" is not a directory'.format(root))

        self._root = root

        self._pcb_paths = {}
        for p in [os.path.join(root, f) for f in os.listdir(root) if f.startswith('pcb') and os.path.isdir(os.path.join(root, f))]:
            id = int(os.path.splitext(os.path.basename(p))[0][3:])
            self._pcb_paths[id] = p

    def num_pcbs(self):
        '''
        Returns the number of PCBs in the dataset.
        '''

        return len(self._pcb_paths)

    def pcb_ids(self):
        '''
        Returns a sorted list of IDs of all PCBs in the dataset.
        '''

        return sorted(list(self._pcb_paths.keys()))

    def pcb(self, id, scale=1):
        '''
        Returns the PCB with the given ID as a PCB object.
        id: PCB id (see pcb_ids()).
        scale: scale factor (1 = original size).
        '''

        if id not in self._pcb_paths:
            raise Exception('Unknown PCB ID')

        return PCB(self._pcb_paths[id], scale)

    def pcbs(self, scale=1):
        '''
        Generator function for returning all PCBs in the dataset as PCB objects.
        scale: scale factor (1 = original size).
        '''

        for id in self._pcb_paths:
            yield PCB(self._pcb_paths[id], scale)


# a simple visualizer to demonstrate the API

if __name__ == "__main__":
    # parse args

    def minmax(s):
        try:
            mi, ma = map(float, s.split(','))
            return mi, ma
        except:
            raise argparse.ArgumentTypeError('min/max syntax is min,max')

    parser = argparse.ArgumentParser(description='Visualize a dataset image')
    parser.add_argument('--root', type=str, dest='root', required=True, help='Path to the dataset')
    parser.add_argument('--pcb', type=int, dest='pcb', default=1, help='ID of the PCB to show')
    parser.add_argument('--rec', type=int, dest='rec', default=1, help='Recording to show')
    parser.add_argument('--scale', type=float, dest='scale', default=1, help='Scale factor')
    parser.add_argument('--icsz', type=minmax, dest='icsz', default='0,0', help='(min, max) size of returned ICs in cm^2 (0 = no restriction)')
    parser.add_argument('--icas', type=minmax, dest='icas', default='0,0', help='(min, max) aspect ratio of returned ICs (0 = no restriction)')
    args = parser.parse_args()

    # load data and show it

    db = PCBDataset(args.root)
    print('Dataset contains images of {} PCBs'.format(db.num_pcbs()))
    print(' {}'.format(db.pcb_ids()))

    pcb = db.pcb(args.pcb, args.scale)
    print('Loaded PCB {}, available recordings: {}'.format(args.pcb, pcb.recordings()))

    ics = pcb.ics(args.rec, True, args.icsz, args.icas)
    print('PCB contains {} ICs'.format(len(ics)))

    img = pcb.image_masked(args.rec)

    for ic in ics:
        bp = cv2.cv.BoxPoints(ic.rect)
        bp = np.int0(bp)
        cv2.drawContours(img, [bp], 0, (0, 255, 0), 2)

    cv2.imshow('PCB', img)
    cv2.waitKey(0)
